teps/0022-trigger-immutable-input.md

```diff
@@ -0,0 +1,409 @@
+be merged together in the `interceptors` field. In cases where the response
+contains a key already present in the payload, the most recent response value
+will replace the existing value.
+
+Webhook interceptors may no longer modify headers as part of the interceptor
+chain.
+
+##### Example
+
+Given a webhook interceptor named `foo` that returns the following response:
+
+```json
+{
+  "bar": "baz"
+}
+```
+
+We expect the next request body to the following interceptor to look like:
+
+```json
+{
+  "headers": {...},
+  "body": {...},
+  "interceptors": {
+    "bar": "baz"
+  }
+}
+```
+
+### CEL Interceptors and TriggerBindings
+
+While there will be no syntax change to CEL Interceptors or Trigger Bindings,
+users will need to be aware that custom fields that might have been embeded may
+have moved to the `interceptors` field.
+
+CEL overlay keys will now be rooted from the `interceptors` field. e.g. given
+the existing example:
+
+```yaml
+- key: body.pull_request.head.short_sha
+  expression: "truncate(body.pull_request.head.sha, 7)"
+```
+
+This will write the data into `interceptors.body.pull_request.head.short_sha`.
+It is recommended that users use short keys instead e.g.
+
+```yaml
+- key: short_sha
+  expression: "truncate(body.pull_request.head.sha, 7)"
+```
+
+## Design Details
+
+### Inteceptor Interface
+
+We plan to follow changes to the Interceptor interface as proposed by the
+[Pluggable Interceptors Design Doc](https://docs.google.com/document/d/1zIG295nyWonCXhb8XcOC41q4YVP9o2Hgg8uAoL4NdPA/edit#).
+
+```go
+interface Interceptor {
+    ExecuteTrigger(req *InterceptorRequest) (*InterceptorResponse, error)
+}
+
+type InterceptorRequest struct {
+  Headers http.Headers
+  Body json.RawMessage
+  Interceptors map[string]interface{}
+}
+
+type InterceptorResponse struct {
+  Body map[string]interface{}
+}
+```
+
+This proposal will not cover adding the additional Trigger / status fields
+included in the Pluggable Interceptors proposal, but is compatible with those
+changes.
+
+## Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy.  Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+-->
+
+We should be able to leverage the existing inteceptor test infrastructure - this
+is a change in behavior to an already existing feature. We should make sure to
+test both new and legacy paths.
+
+## Drawbacks
+
+<!--
+Why should this TEP _not_ be implemented?
+-->
+
+```
+<<[UNRESOLVED wlynch ]>>
+TODO based on feedback
+<<[/UNRESOLVED]>>
+```
```

Comment:
one potential drawback that comes to mind: can't as easily use interceptors to filter out sensitive data? e.g. if you could imagine wanting to pass the entire body of an event to a resulting Run

i suppose you can still do this, but you need to do it by duplicating all the data into a new field?

Reply:
ooh, great point! Added (though I don't think we want to support this).